File: 15. 3Sum/3Sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) 
    {
        //threeSumSet is a set list of integers that we can use to store the triplets
        Set<List<Integer>> threeSumSet = new HashSet<>();
        //First sort the array
        Arrays.sort(nums);
        //We will do a 2 way sweep with three pointers
        //i will be the index we are looking at, j will look at the elemenet next to i, and k will look at the element at the end of the array
        for (int i = 0; i < nums.length - 2; i++)
        {
            int j = i + 1;
            int k = nums.length - 1;
            
            while (j < k)
            {
                int sum = nums[i] + nums[j] + nums[k];
                //If sum == 0, we add that triplet to threeSumSet
                if (sum == 0)
                {
                    threeSumSet.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    j++;
                    k--;
                }
                //If sum > 0; we decrease the increment of k
                else if (sum > 0)
                {
                    k--;
                }
                //If sum < 0; we increase the increment of j
                else if (sum < 0)
                {
                    j++;
                }
                //We do not need to worry about duplicate triplets because adding into a set automatically checks the lists to make sure there are no duplicates
            }
        }
        //convert the set to an ArrayList and return
        return new ArrayList<>(threeSumSet);
    }
}

File: 15. 3Sum/3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        #"threeSumSet" is a list of integer that we can use to store triplets
        threeSumSet = []
        #First sort the list nums
        nums.sort()
        #We will do a 2 way sweep with three pointers
        #"i" will be the index we are looking at, "j" will look at the element next to "i", and "k" will look at the elment at the end of the array
        for i in xrange(len(nums) - 2):
            #This if statement ensures that we do not insert duplicates into our list "threeSumSet"
            if i > 0 and nums[i] == nums[i-1]:
                continue
            j = i + 1
            k = len(nums) - 1
            while j < k:
                sum = nums[i] + nums[j] + nums[k]
                #If "sum" is 0, that means we have found a triplet and should accordingly add the triplet to the list "threeSumSet"
                if sum == 0:
                    threeSumSet.append((nums[i], nums[j], nums[k]))
                    #The bottom two while loops ensure that we do not enter a duplicate triplet
                    while j < k and nums[j] == nums[j+1]:
                        j += 1
                    while j < k and nums[k] == nums[k-1]:
                        k -= 1
                    j += 1
                    k -= 1
                #If the sum is greater than 0, we decrease the increment of k
                elif sum > 0:
                    k -= 1
                #If the sum is less than 0, we increase the increment of j
                elif sum < 0:
                    j += 1
        #return
        return threeSumSet